### PR TITLE
3032 Move to yarn

### PR DESCRIPTION
**Pull request by rmhowe**
# Description

For reasons discussed on the issue, we're moving from npm to yarn.

A few points worth noting:
- One major complication in this move was that based on our dependencies, yarn was installing 2 versions of jQuery (where npm was only installing one). This broke the app to the point where it wouldn't load. Npm resolved this by only using the lower of the 2 versions, and yarn allows setting deep dependencies so that's what we've done here, however this may not be a sustainable solution if these packages bump their dependencies and explicitly require later jQuery versions. As it stands this would be a blocking issue for moving to yarn.
- We all need to be on the same version of yarn for the `yarn.lock` file to work correctly and be the same for all of us. In my opinion the best way to achieve this is using the `engines` field in `package.json` to explicitly set a yarn version for use with `medic-webapp`. This has some complications though as in the next point
- The `engines` field is used inconsistently in npm packages. In theory it should be used to determine which versions of node/npm a particular package requires, however some packages we require (eg. `angular-pouchdb`) have set their node versions very explicitly, and if you are outside of this then yarn will throw an error such as the one below. This means that until packages correctly use the `engines` field, yarn needs to be called using the `--ignore-engines` flag. This is a slight pain and also means our own `engines` field (which tells you which version of yarn is required) will be ignored as well, but seems necessary.
```
error angular-pouchdb@5.0.1: The engine "node" is incompatible with this module. Expected version ">=4 <5".
```

cc @garethbowen @SCdF @alxndrsn @dianabarsan @vimemo 

medic/medic-webapp#3032

# Review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- [ ] Documented: Announced in Changes.md in plain English. Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

### CODE REVIEW

**Comment by garethbowen** [comment]
> ... if these packages bump their dependencies and explicitly require later jQuery versions ... this would be a blocking issue for moving to yarn.

As it stands this would break in npm too. IIRC the problem is the jquery plugins are attached to whatever instance is in the browser. If a module uses a different instance then the plugins are not available. Either way we want to force all dependencies to use the same jquery.

> We all need to be on the same version of yarn...

We could write some manual task to check the local version of yarn matches if necessary...

> The engines field is used inconsistently in npm packages.

Is angular-pouch the only dependency that's failing this check? Fortunately the author is willing to accept contributions - we recently submitted a patch to relax the pouchdb requirement for example. Probably the fix here is just to remove the `engines` property in angular-pouchdb because it doesn't make sense in the browser anyway...

File: Gruntfile.js [hunk 1]
```diff
@@ -586,9 +587,9 @@ module.exports = function(grunt) {
   });
 
   // Build tasks
-  grunt.registerTask('mmnpm', 'Update and patch npm dependencies', [
+  grunt.registerTask('mmyarn', 'Update and patch dependencies', [
```
Comment (garethbowen) [inline Gruntfile.js]: Maybe we should rename this to something generic and more descriptive, eg: `install_dependencies`?

File: Gruntfile.js [hunk 2]
```diff
@@ -740,7 +741,7 @@ module.exports = function(grunt) {
     'jshint',
   ]);
 
-  grunt.registerTask('dev-webapp-no-npm', 'Build and deploy the webapp for dev, without reinstalling dependencies.', [
+  grunt.registerTask('dev-webapp-no-yarn', 'Build and deploy the webapp for dev, without reinstalling dependencies.', [
```
Comment (garethbowen) [inline Gruntfile.js]: Same here? `dev-webapp-no-dependencies`?

File: package.json [hunk 3]
```diff
@@ -114,5 +118,8 @@
       "path": "./static/dist/inbox.css",
       "maxSize": "70 kB"
     }
-  ]
+  ],
+  "resolutions": {
+    "**/jquery": "3.2.x"
+  }
```
Comment (garethbowen) [inline package.json]: SWEET!

File: Gruntfile.js (hunk 1 above)
Comment (rmhowe) [reply Gruntfile.js]: Yep makes sense

**Comment by rmhowe** [comment]
@garethbowen could you take another look at this? Mainly just the travis file changes and moving the starting of webdriver into grunt rather than the travis file.

> Is angular-pouch the only dependency that's failing this check? Fortunately the author is willing to accept contributions - we recently submitted a patch to relax the pouchdb requirement for example. Probably the fix here is just to remove the engines property in angular-pouchdb because it doesn't make sense in the browser anyway...

I'm not sure but I have a feeling `angular-pouch` is just the first package that fails the engine check and there'll be more further down the line, regardless that sounds like a good option to try to make PRs against the repos for these packages.

**Pull request merged**